### ISSUE TITLE
Code insights: Move from extension-based data fetching to built-in Apollo fetchers

### ISSUE THREAD

**Issue by vovakulikov** [sourcegraph/sourcegraph#24236]
## Background

At the moment our extension-based charts (line chart and pie chart) are working vie extension API. At the same time, we have live-preview charts (with almost the same functionality) for the live preview at the creation UI page. Data fetching logic for the Live preview charts were copied from the extensions code-bases for the 3.28 release. Since that moment we have same logic for data-fetching in the two places 

- Sourcegraph App - [line-chart preview](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/insights/core/backend/types.ts?L119:5&subtree=true) and [pie-chart preview](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/insights/core/backend/types.ts?L126&subtree=true)
- Extension codebases - [Search Insight (line-chart)](https://github.com/sourcegraph/sourcegraph-search-insights/blob/master/src/search-insights.ts), [Code Stats Insight (pie-chart)](https://github.com/sourcegraph/sourcegraph-code-stats-insights/blob/master/src/code-stats-insights.ts)

This station increases complexity and puts us in a position to implement functionality twice for live preview and extension-based chart on the dashboard page.

In order to avoid code duplication in live-preview and insight cards charts on the dashboard page, we should use built-in data-fetchers within sourcegraph app and move away from the extension API for this two.

This refactoring also unlocks us to adopt apollo for chart data fetching which opens for us a lot of possibilities (like cache strategies and request batching) to improve our loading strategy and have a data cache for insights data. 

## Tech details 

- At the moment we have a support of one-by-one insight fetching logic [via our own hook useParallelRequest](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/insights/hooks/use-parallel-requests/use-parallel-request.ts?L136:36&subtree=true). We probably want to preserve this functionality for the insight loading even if we get benefits from the apollo solution. This logic might be preserved by a custom apollo link. You can find proof of concept of this kind of one-by-one logic thing [here](https://github.com/vovakulikov/apollo-parallel/blob/main/src/parallel-link.ts)

- We also should keep in mind that we have Search and Tree pages that can render insights grid too. At the moment they use a stateless approach to render insights (consumer/page is supposed to fetch data and then passes that data to code insight visual components `StaticInsightsViewGrid`). We need to revisit loading on these pages and probably align the insight loading API with the dashboard page.

## Migration plan

- [x] Migrate data-fetchers from the extensions insight codebases to the main sg app (in this step for the dashboard page only). By this, we get rid of extension API on the dashboard page that can solve this issue with extension insight activation problem https://github.com/sourcegraph/sourcegraph/issues/24262
This change was completed by this PR https://github.com/sourcegraph/sourcegraph/pull/24559
- [x] Revisit the loading system at consumer insights pages (Home and Tree pages) and align the insight loading approach with the Dashboard page loading API.
Also, this change was closed by this PR https://github.com/sourcegraph/sourcegraph/pull/24559
- [x] Implement Apollo parallel request link to be able to preserve current fetch logic (load maximum two insights at the time - in parallel)
- [ ] Migrate data-fetchers from the code insight-specific API context to the apollo hooks approach.
- [ ] **[Streatch goal]** Consider refactoring insight one insight by one insight loading logic to the Apollo links system instead of useParallelLoading hook for built-in type of insights.

**Comment by felixfbecker** [comment]
@vovakulikov this issue is still in the last iteration milestone, what is the status of it? Do you want to make a case to include it in the next iteration, or to backlog it?

**Comment by vovakulikov** [comment]
@felixfbecker I think it makes sense to include this issue in our current (2021-09-13) iteration since we already have done a lot of work to be able to finish this migration. We've got some problems to finish that according to our initial plan for this (multiple requests per insight is a problem at the moment) but it's solvable and I think this ticket requires ~2 days to finish that completely (by this I mean adopt Apollo cache for the dashboard page). Let's back this one to our current iteration and mark this as a stretch goal.

**Comment by felixfbecker** [comment]
To clarify, you estimate 2 remaining days to finish what part of this issue – all the unticked checkboxes above?

**Comment by vovakulikov** [comment]
@felixfbecker That's correct all unchecked above. I would consider making those changes next week since this week is all about preparing our public beta things.
